### PR TITLE
Rhb 24

### PR DESCRIPTION
**Pull request by EkaterinaKaonashi**

### CODE REVIEW

File: DevEdu.API/DevEdu.API/Controllers/CourseController.cs [hunk 1]
```diff
@@ -146,6 +146,26 @@ public List<CourseTopicOutputModel> SelectAllTopicsByCourseId(int courseId)
             return _mapper.Map<List<CourseTopicOutputModel>>(list);
             
         }
+        // api/course/{courseId}
+        [HttpPut("{courseId}/update")]
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.API/Controllers/CourseController.cs:150]: please rename to "{courseId}/program" (and the comment on 149 line as well)

File: DevEdu.API/DevEdu.API/Controllers/CourseController.cs [hunk 2]
```diff
@@ -146,6 +146,26 @@ public List<CourseTopicOutputModel> SelectAllTopicsByCourseId(int courseId)
...
+        [ProducesResponseType(typeof(string), StatusCodes.Status200OK)]
+        public string UpdateCourseTopicsByCourseId(int courseId,  [FromBody] List<CourseTopicUpdateInputModel> topics)
+        {
+            var list = _mapper.Map<List<CourseTopicDto>>(topics);
+            _courseService.UpdateCourseTopicsByCourseId(courseId, list);
+            return "updated";
+
+        }
+        // api/course/{courseId}/topics-delete
+        [HttpDelete("{courseId}/topics-delete")]
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.API/Controllers/CourseController.cs:161]: do we really need to delete all? 😱😱😱

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs [hunk 3]
```diff
@@ -44,5 +47,49 @@ public List<CourseTopicDto> SelectAllTopicsByCourseId(int courseId)
...
+            }
+        }
+        private void CheckUniquenessPositions(List<CourseTopicDto> topics)
+        {
+            if (topics.GroupBy(n => n.Position).Any(c => c.Count() > 1))
+            {
+                throw new Exception("the same positions of topics in the course");
+            }
+        }
+        private CourseDto SetIdToCourseDto(int courseId, CourseDto dto)
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/CourseService.cs:90]: please remove this method

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs [hunk 4]
```diff
@@ -44,5 +47,49 @@ public List<CourseTopicDto> SelectAllTopicsByCourseId(int courseId)
...
+            {
+                AddTopicToCourse(courseId, topic.Topic.Id, topic);
+            }
+        }
+        private void CheckUniquenessPositions(List<CourseTopicDto> topics)
+        {
+            if (topics.GroupBy(n => n.Position).Any(c => c.Count() > 1))
+            {
+                throw new Exception("the same positions of topics in the course");
+            }
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/CourseService.cs:88]: nice approach)
maybe it's better to add such checking for TopicIds?

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs [hunk 5]
```diff
@@ -44,5 +47,49 @@ public List<CourseTopicDto> SelectAllTopicsByCourseId(int courseId)
             var list = _courseRepository.SelectAllTopicsByCourseId(courseId);
             return list;
         }
+        public void UpdateCourseTopicsByCourseId(int courseId, List<CourseTopicDto> topics)
+        {
+            CheckUniquenessPositions(topics);
+            var topicsInDatabase = _courseRepository.SelectAllTopicsByCourseId(courseId);
+            if (((topics.Count > topicsInDatabase.Count) || (topics.Count < topicsInDatabase.Count)) && topicsInDatabase.Count!=0)
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/CourseService.cs:54]: possible NullReferenceException in topics.Count and in topicsInDatabase.Count

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs [hunk 6]
```diff
@@ -44,5 +47,49 @@ public List<CourseTopicDto> SelectAllTopicsByCourseId(int courseId)
...
+            {
+                foreach (var topic in topics)
+                {
+                    topic.Course = SetIdToCourseDto(courseId, topic.Course);
+                }
+                _courseRepository.UpdateCourseTopicsByCourseId(courseId, topics);
+            }
+        }
+        public void DeleteAllTopicsByCourseId(int courseId)
+        {
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/CourseService.cs:73]: please use the "short" version of method (with lambda)

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs [hunk 7]
```diff
@@ -44,5 +47,49 @@ public List<CourseTopicDto> SelectAllTopicsByCourseId(int courseId)
...
+            _courseRepository.DeleteAllTopicsByCourseId(courseId);
+        }
+        private void AddTopicsFromNewList(int courseId, List<CourseTopicDto> topics)
+        {
+            foreach (var topic in topics)
+            {
+                AddTopicToCourse(courseId, topic.Topic.Id, topic);
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.Business/Services/CourseService.cs:80]: we might have ~50 topics in a course
and it means that we'll send 50 requests to DB here
I think it's not the best solution(

File: DevEdu.API/DevEdu.DAL/Repositories/CourseRepository.cs [hunk 8]
```diff
@@ -166,5 +168,29 @@ public List<CourseTopicDto> SelectAllTopicsByCourseId(int courseId)
                 )
                 .ToList();
         }
+        public void UpdateCourseTopicsByCourseId(int courseId, List<CourseTopicDto> topics)
+        {
+            foreach (var topic in topics)
+            {
+                _connection.Query(
+                    _updateCourseTopicsProcedure,
```
Comment (Negodyaj) [inline DevEdu.API/DevEdu.DAL/Repositories/CourseRepository.cs:176]: we might have ~50 topics in a course
and it means that we'll send 50 requests to DB here
I think it's not the best solution(

File: DevEdu.API/DevEdu.API/Controllers/CourseController.cs (hunk 1 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.API/Controllers/CourseController.cs:150]: fixed

File: DevEdu.API/DevEdu.API/Controllers/CourseController.cs (hunk 2 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.API/Controllers/CourseController.cs:161]: why not? Maybe for someone will be easier to delete all topics than delete one by one. If you say to delete then I'll delete.

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs (hunk 3 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.Business/Services/CourseService.cs:90]: fixed

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs (hunk 4 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.Business/Services/CourseService.cs:88]: added, but maybe that's not what you wanted

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs (hunk 5 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.Business/Services/CourseService.cs:54]: fixed

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs (hunk 6 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.Business/Services/CourseService.cs:73]: fixed

File: DevEdu.API/DevEdu.Business/Services/CourseService.cs (hunk 7 above)
Comment (EkaterinaKaonashi) [reply DevEdu.API/DevEdu.Business/Services/CourseService.cs:80]: fixed

**Pull request merged**